invoices/invoice_expiry_watcher.go

```diff
@@ -0,0 +1,193 @@
+package invoices
+
+import (
+	"fmt"
+	"sync"
+	"time"
+
+	"github.com/lightningnetwork/lnd/channeldb"
+	"github.com/lightningnetwork/lnd/clock"
+	"github.com/lightningnetwork/lnd/lntypes"
+	"github.com/lightningnetwork/lnd/queue"
+	"github.com/lightningnetwork/lnd/zpay32"
+)
+
+// invoiceExpiry holds and invoice's payment hash and its expiry. This
+// is used to order invoices by their expiry for cancellation.
+type invoiceExpiry struct {
+	PaymentHash lntypes.Hash
+	Expiry      time.Time
+}
+
+// Less implements PriorityQueueItem.Less such that the top item in the
+// priorty queue will be the one with that expires next.
+func (e invoiceExpiry) Less(other queue.PriorityQueueItem) bool {
+	return e.Expiry.Before(other.(*invoiceExpiry).Expiry)
+}
+
+// InvoiceExpiryWatcher handles automatic invoice cancellation of expried
+// invoices.
+// Upon start InvoiceExpiryWatcher will subscribe to all invoice events and
+// will add all active (not yet settled and not yet cancelled) invoices to its
+// watcing queue.
+// When a new invoice is added to the InvoiceRegistry, InvoiceExpiryWatcher
+// will add that to the watching queue as well.
+// If any of the watched invoices expire, they'll be removed from the watching
+// queue and will be cancelled through InvoiceRegistry.CancelInvoice.
+type InvoiceExpiryWatcher struct {
+	sync.Mutex
+	started bool
```

Comment:
This deviates from the customary `started/stopped` atomic variables we use elsewhere in the codebase.

Reply:
I prefer using the mutex as that way either `Start()` or `Stop()` runs.

Reply:
See also discussion below. I'm fine with changing this but believe it's better (see comments).